drivers/timer/apic_tsc.c

```diff
@@ -0,0 +1,189 @@
+	uint8_t unused0  : 8;
+	uint8_t masked   : 1;
+	enum { ONE_SHOT, PERIODIC, TSC_DEADLINE } mode : 2;
+	uint32_t unused2 : 13;
+};
+
+static struct k_spinlock lock;
+static uint64_t last_announce;
+static union { uint32_t val; struct apic_timer_lvt lvt; } lvt_reg;
+
+static ALWAYS_INLINE uint64_t rdtsc(void)
+{
+	uint32_t hi, lo;
+
+	__asm__ volatile("rdtsc" : "=d"(hi), "=a"(lo));
+	return lo + (((uint64_t)hi) << 32);
+}
+
+static void isr(const void *arg)
+{
+	ARG_UNUSED(arg);
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	uint32_t ticks = (rdtsc() - last_announce) / CYC_PER_TICK;
+
+	last_announce += ticks * CYC_PER_TICK;
+	k_spin_unlock(&lock, key);
+	sys_clock_announce(ticks);
+
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		sys_clock_set_timeout(1, false);
+	}
+}
+
+static inline void wrmsr(int32_t msr, uint64_t val)
+{
+	uint32_t hi = (uint32_t) (val >> 32);
+	uint32_t lo = (uint32_t) val;
+
+	__asm__ volatile("wrmsr" :: "d"(hi), "a"(lo), "c"(msr));
+}
+
+void sys_clock_set_timeout(int32_t ticks, bool idle)
+{
+	ARG_UNUSED(idle);
+
+	uint64_t now = rdtsc();
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	uint64_t expires = now + MAX(ticks - 1, 0) * CYC_PER_TICK;
+
+	expires = last_announce + (((expires - last_announce + CYC_PER_TICK - 1)
+				    / CYC_PER_TICK) * CYC_PER_TICK);
+
+	/* The second condition is to catch the wraparound.
+	 * Interpreted strictly, the IA SDM description of the
+	 * TSC_DEADLINE MSR implies that it will trigger an immediate
+	 * interrupt if we try to set an expiration across the 64 bit
+	 * rollover.  Unfortunately there's no way to test that as on
+	 * real hardware it requires more than a century of uptime,
+	 * but this is cheap and safe.
+	 */
+	if (ticks == K_TICKS_FOREVER || expires < last_announce) {
+		expires = UINT64_MAX;
+	}
+
+	wrmsr(IA32_TSC_DEADLINE_MSR, expires);
+	k_spin_unlock(&lock, key);
+}
+
+uint32_t sys_clock_elapsed(void)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	uint32_t ret = (rdtsc() - last_announce) / CYC_PER_TICK;
+
+	k_spin_unlock(&lock, key);
+	return ret;
+}
+
+uint32_t sys_clock_cycle_get_32(void)
+{
+	return (uint32_t) rdtsc();
+}
+
+static inline uint32_t timer_irq(void)
+{
+	/* The Zephyr APIC API is... idiosyncratic.  The timer is a
+	 * "local vector table" interrupt.  These aren't system IRQs
+	 * presented to the IO-APIC, they're indices into a register
+	 * array in the local APIC.  By Zephyr convention they come
+	 * after all the external IO-APIC interrupts, but that number
+	 * changes depending on device configuration so we have to
+	 * fetch it at runtime.  The timer happens to be the first
+	 * entry in the table.
+	 */
+	return z_loapic_irq_base();
+}
+
+void smp_timer_init(void)
+{
+	/* Copy the LVT configuration from CPU0, because IRQ_CONNECT()
+	 * doesn't know how to manage LVT interrupts for anything
+	 * other than the calling/initial CPU.  Same fence needed to
+	 * prevent later MSR writes from reordering before the APIC
+	 * configuration write.
```

Comment:
Note: this is one of the glitches with SMP support in the older APIC driver, and really it's a hole in our APIC support.  The timer driver shouldn't be responsible for this, new CPUs should be automatically set up with the correctly mapped and configured interrupts by the arch layer.  But it's not like it's difficult to do here.